db: Avoid memtable flush latency on schema merge

tgrabiec opened the pull request:
Currently, applying schema mutations involves flushing all schema
tables so that on restart commit log replay is performed on top of
latest schema (for correctness). The downside is that schema merge is
very sensitive to fdatasync latency. Flushing a single memtable
involves many syncs, and we flush several of them. It was observed to
take as long as 30 seconds on GCE disks under some conditions.

This patch changes the schema merge to rely on a separate commit log
to replay the mutations on restart. This way it doesn't have to wait
for memtables to be flushed. It has to wait for the commitlog to be
synced, but this cost is well amortized.

We put the mutations into a separate commit log so that schema can be
recovered before replaying user mutations. This is necessary because
regular writes have a dependency on schema version, and replaying on
top of latest schema satisfies all dependencies. Without this, we
could get loss of writes if we replay a write which depends on the
latest schema on top of old schema.

Also, if we have a separate commit log for schema we can delay schema
parsing for after the replay and avoid complexity of recognizing
schema transactions in the log and invoking the schema merge logic.

Fixes #8272
Fixes #8309
Fixes #1459

avikivity commented on replica/database.cc:
```diff
@@ -1756,6 +1756,50 @@ future<> database::apply_with_commitlog(column_family& cf, const mutation& m, db
     }
 }
 
+future<> database::apply(const std::vector<frozen_mutation>& muts, db::timeout_clock::time_point timeout) {
```
Let's call it apply_in_memory() (like the existing one) so people aren't tempted to think it is durable.

avikivity replied:
yes, doh

avikivity commented on replica/database.cc:
```diff
@@ -1756,6 +1756,50 @@ future<> database::apply_with_commitlog(column_family& cf, const mutation& m, db
+    // FIXME: Memtable application is not atomic so reads may observe mutations partially applied until restart.
+    for (auto i = 0; i < muts.size(); ++i) {
+        auto&& cf = find_column_family(muts[i].column_family_id());
+        auto s = local_schema_registry().get(muts[i].schema_version());
```
Let's move these to a separate pass so a failure here doesn't end with inconsistent state (like the FIXME).

What would a fix look like? Maybe creating a new MVCC version but not linking it in, and then the final pass links in pre-prepared versions into the partition.

tgrabiec replied:
Yeah, something like that. I can work on it if we agree on the approach, incrementally. It's not worse than the current state so it's not a blocker IMO.

avikivity commented on replica/database.cc:
```diff
@@ -1756,6 +1756,50 @@ future<> database::apply_with_commitlog(column_family& cf, const mutation& m, db
+    for (auto i = 0; i < muts.size(); ++i) {
+        auto&& cf = find_column_family(muts[i].column_family_id());
+        auto s = local_schema_registry().get(muts[i].schema_version());
+        co_await apply_in_memory(muts[i], s, std::move(handles[i]), timeout);
```
The timeout is also a source of partial applies. I guess for the schema it doesn't matter, due to infinite timeout.

avikivity commented on db/schema_tables.cc:
```diff
@@ -156,7 +156,7 @@ static future<user_types_to_drop> merge_types(distributed<service::storage_proxy
 static future<> merge_aggregates(distributed<service::storage_proxy>& proxy, schema_result before, schema_result after);
 
-static future<> do_merge_schema(distributed<service::storage_proxy>&, std::vector<mutation>, bool do_flush);
+static future<> do_merge_schema(distributed<service::storage_proxy>&, std::vector<mutation>, bool bypass_commitlog);
```
Doesn't that invert the sense? previously, true == safe. Now, true == unsafe.

Safer to keep the sense in case a caller is missed.

tgrabiec replied:
True, will change

gleb-cloudius commented on replica/database.hh:
```diff
@@ -1499,6 +1499,13 @@ public:
     future<> apply(schema_ptr, const frozen_mutation&, tracing::trace_state_ptr tr_state, db::commitlog_force_sync sync, db::timeout_clock::time_point timeout);
+    // Apply mutations atomically.
+    // On restart, either all mutations will be replayed or none of them.
+    // All mutations must belong to the same commitlog domain.
```
Can you remind me please what is "commit log domain"?

tgrabiec replied:
It's `db::commitlog` instance. Before this patch, there was a single domain used by all writes. Now we have two domains, for system_schema and regular.

gleb-cloudius replied:
You mean after the next patch?

tgrabiec replied:
Yes.

avikivity commented on replica/database.cc:
```diff
@@ -1756,6 +1756,50 @@ future<> database::apply_with_commitlog(column_family& cf, const mutation& m, db
+            }
+            shard = m_shard;
+        } else if (*shard != m_shard) {
+            throw_with_backtrace<std::runtime_error>("Cannot apply atomically across shards");
```
on_internal_error throws with a backtrace

avikivity commented on main.cc:
```diff
@@ -1029,9 +1029,6 @@ To start the scylla server proper, simply invoke as: scylla server (or just scyl
-            supervisor::notify("loading non-system sstables");
-            replica::distributed_loader::init_non_system_keyspaces(db, proxy, sys_ks).get();
-
             supervisor::notify("starting view update generator");
```
Is it safe to move this after the components below?

Moving it after enable_features_on_startup is probably an improvement.

tgrabiec replied:
I'm not 100% sure about view_update_generator. To be safe, I'll move it after init_non_system_keyspaces(). 

Other things seem safe to reorder.

avikivity commented on main.cc:
```diff
@@ -1047,6 +1044,27 @@ To start the scylla server proper, simply invoke as: scylla server (or just scyl
+                    }).get();
+                    supervisor::notify("replaying schema commit log - removing old commitlog segments");
+                    //FIXME: discarded future
+                    (void)sch_cl->delete_segments(std::move(paths));
```
Why not wait?

avikivity replied:
And don't we need to reload the schema?

tgrabiec replied:
This mirrors the existing logic for regular replay. I'm not sure why it doesn't wait.

We don't need to reload the schema because the schema is loaded only after replay.

avikivity replied:
ok & ok

avikivity commented on replica/database.cc:
```diff
@@ -801,21 +801,43 @@ future<> database::parse_system_tables(distributed<service::storage_proxy>& prox
+        db::commitlog::config c;
+        c.commit_log_location = _cfg.commitlog_directory();
+        c.fname_prefix = db::schema_tables::COMMITLOG_FILENAME_PREFIX;
+        c.metrics_category_name = "sch-commitlog";
```
Let's make the name less obscure.

tgrabiec replied:
```
c.metrics_category_name = "schema-commitlog";
```

avikivity commented on replica/database.cc:
```diff
@@ -801,21 +801,43 @@ future<> database::parse_system_tables(distributed<service::storage_proxy>& prox
+        c.commitlog_total_space_in_mb = 10 >> 20;
+        c.commitlog_segment_size_in_mb = _cfg.commitlog_segment_size_in_mb();
+        c.commitlog_sync_period_in_ms = _cfg.commitlog_sync_period_in_ms();
+        c.mode = db::commitlog::sync_mode::BATCH;
```
Important

avikivity commented on replica/database.cc:
```diff
@@ -863,7 +885,8 @@ void database::add_column_family(keyspace& ks, schema_ptr schema, column_family:
     lw_shared_ptr<column_family> cf;
     if (cfg.enable_commitlog && _commitlog) {
-       cf = make_lw_shared<column_family>(schema, std::move(cfg), *_commitlog, *_compaction_manager, *_cl_stats, _row_cache_tracker);
+        db::commitlog& cl = schema->ks_name() == "system_schema" ? *_schema_commitlog : *_commitlog;
```
Needs to depend on a cluster feature, so we can downgrade.

avikivity replied:
(the decision whether to flush too)

avikivity commented:
Very good, but also frightening. We'll need good tests for this.

gleb-cloudius commented on db/schema_tables.cc:
```diff
@@ -879,9 +879,10 @@ future<> update_schema_version_and_announce(sharded<db::system_keyspace>& sys_ks
-        bool flush_schema = proxy.local().get_db().local().get_config().flush_schema_tables_after_modification();
-        co_await do_merge_schema(proxy, std::move(mutations), flush_schema);
-        co_await update_schema_version_and_announce(sys_ks, proxy, feat.cluster_schema_features());
+        co_await smp::submit_to(0, [&] () mutable -> future<> {
```
Isn't it better better to jump to shard zero before calling `with_merge_lock`? Otherwise you jump there 3 times.

tgrabiec replied:
It is slightly better, yes.

gleb-cloudius commented on replica/database.cc:
```diff
@@ -801,21 +801,43 @@ future<> database::parse_system_tables(distributed<service::storage_proxy>& prox
 database::init_commitlog() {
     if (_commitlog) {
-        return make_ready_future<>();
+        co_return co_await make_ready_future<>();
```
Isn't this just `co_return`?

gleb-cloudius replied:
That is if you write `co_return make_ready_future<>();`, no? What about `co_return;`?

tgrabiec replied:
Just `co_return;` works, thanks

tgrabiec commented:
In V2:

 - Added code to update metrics around multi-key apply

 - Added missing handling of schema commitlog shutdown
 
 - Fixed passing of unfrozen mutations across shards in merge_schema()

 - Use no abbreviations for `metrics_category_name`
 
 - Start view update generator after init_non_system_keyspaces(), like before

 - Improved logging in do_apply_many() 
 
 - Improved logging on error during commitlog replay
 
 - Use on_internal_error() in do_apply_many()
  
 - Added cluster feature to guard the new behavior - only after full upgrade and rolling restart nodes will use the schema commitlog
  
    The commitlog domain is switched only when all nodes are upgraded, and
    only after new node is restarted. This is so that we don't have to add
    risky code to deal with hot-switching of the commitlog domain. Cold
    switching is safer. This means that after upgrade there is a need for
    yet another rolling restart round. 

 - Disabled support for truncation of schema tables to prevent old replay positions to be interpreted in the context of new commitlog domain

    One complication with switching commitlog domains is that replay_position markers are
    commitlog-domain specific and cannot cross domains. They are recorded
    in various places which survive node restart: sstables are annotated
    with the maximum replay position, and they are present inside
    truncation records. The former annotation is used by "truncate"
    operation to drop sstables. To prevent old replay positions from being
    interpreted in the context in the new schema commitlog domain, the
    change refuses to boot if there are truncation records, and also
    prohibits truncation of schema tables.
  
  - System table initialization is split into two phases: non-schema tables and schema tables.
  
    The boot sequence needs to know whether the cluster feature associated
    with this change was enabled on all nodes. Features are stored in
    system.scylla_local. Because we need to read it before initializing
    schema tables, the initialization of tables now has to be split into
    two phases. The first phase initializes all system tables except
    schema tables, and later we initialize schema tables, after reading
    stored cluster features.

 - Tested under stress and kill -9 and hard power failure (no loss of schema changes) 

 - Performance testing:
   
    I reproduced bad behavior locally on my machine with a tired (high latency)
    SSD disk, load driver remote. Under high load, I saw table alter (server-side part) taking
    up to 10 seconds before. After the patch, it takes up to 200 ms (50:1 improvement).
    Without load, it is 300ms vs 50ms.

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/benipeled_test/18/

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/147/

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/146/

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/benipeled_test/19/

kostja commented:
Is group 0 log part of the new commit log? Would you like it to be a part of it?

avikivity commented on db/system_keyspace.cc:
```diff
@@ -1577,7 +1577,10 @@ template <typename T>
     auto type = data_type_for<T>();
-    return qctx->execute_cql(req, type->to_string_impl(data_value(value)), key).discard_result();
+    co_await qctx->execute_cql(req, type->to_string_impl(data_value(value)), key).discard_result();
+    co_await smp::invoke_on_all([] () -> future<> {
```
Let's have a comment explaining this flush so someone doesn't try to remove it in the future (or so we can remove it in ~2 years).

avikivity commented:
How does this interact with Raft?

Should the new commitlog instead be dedicated to all tables managed by group 0? I don't think we want the commitlog for different tables managed by group 0 to be split and replayed out-of-order.

avikivity commented:
I would like to see dtests that demonstrate recovery after a schema change + dependent table change (in extra durable mode where with O_DSYNC things).

tgrabiec commented:
> Is group 0 log part of the new commit log? Would you like it to be a part of it?

> 
> How does this interact with Raft?
> 
> Should the new commitlog instead be dedicated to all tables managed by group 0? I don't think we want the commitlog for different tables managed by group 0 to be split and replayed out-of-order.

I think this change is orthogonal to raft. Raft will work regardless. Commitlog replay restores the state of the state machine. The state machine in this case is mutation-based, so it doesn't matter that mutations are split and reordered, after replay we will end up with the same state. Raft starts only after table commitlogs are replayed, so partial state should not be observed.

avikivity commented:
> > Is group 0 log part of the new commit log? Would you like it to be a part of it?
> 
> > How does this interact with Raft?
> > Should the new commitlog instead be dedicated to all tables managed by group 0? I don't think we want the commitlog for different tables managed by group 0 to be split and replayed out-of-order.
> 
> I think this change is orthogonal to raft. Raft will work regardless. Commitlog replay restores the state of the state machine. The state machine in this case is mutation-based, so it doesn't matter that mutations are split and reordered, after replay we will end up with the same state. Raft starts only after table commitlogs are replayed, so partial state should not be observed.

I see. Still, doesn't it make sense to give preferential treatment for topology as we do for schema changes?

Also, reduce the number of combinations of {is/is not raft group0} x {uses/does not use special commitlog} to just two.

tgrabiec commented:
> > > Is group 0 log part of the new commit log? Would you like it to be a part of it?
> > 
> > 
> > > How does this interact with Raft?
> > > Should the new commitlog instead be dedicated to all tables managed by group 0? I don't think we want the commitlog for different tables managed by group 0 to be split and replayed out-of-order.
> > 
> > 
> > I think this change is orthogonal to raft. Raft will work regardless. Commitlog replay restores the state of the state machine. The state machine in this case is mutation-based, so it doesn't matter that mutations are split and reordered, after replay we will end up with the same state. Raft starts only after table commitlogs are replayed, so partial state should not be observed.
> 
> I see. Still, doesn't it make sense to give preferential treatment for topology as we do for schema changes?

Yes, we could. They are less sensitive than schema changes though. Topology change commands don't need to flush system tables, since they don't need the topology to be available before commitlog replay. Writes to system.token_metadata can still race with user load.

avikivity commented:
> > > > Is group 0 log part of the new commit log? Would you like it to be a part of it?
> > > 
> > > 
> > > > How does this interact with Raft?
> > > > Should the new commitlog instead be dedicated to all tables managed by group 0? I don't think we want the commitlog for different tables managed by group 0 to be split and replayed out-of-order.
> > > 
> > > 
> > > I think this change is orthogonal to raft. Raft will work regardless. Commitlog replay restores the state of the state machine. The state machine in this case is mutation-based, so it doesn't matter that mutations are split and reordered, after replay we will end up with the same state. Raft starts only after table commitlogs are replayed, so partial state should not be observed.
> > 
> > 
> > I see. Still, doesn't it make sense to give preferential treatment for topology as we do for schema changes?
> 
> Yes, we could. They are less sensitive than schema changes though. Topology change commands don't need to flush system tables, since they don't need the topology to be available before commitlog replay. Writes to system.token_metadata can still race with user load.

Note, this is partially due to our design decisions. For example, if regular compaction always performed a cleanup, then we'd need tokens to be fully loaded and up-to-date before starting any compaction of a user table.

Similarly, if we have compaction groups and tablets, then sorting sstables into their tablets requires that we have up-to-date tablet boundaries.

So I think it makes sense to treat topology in the same way as schema, even more than we already do.

kostja commented:
Having thought about it a bit more I still believe that we're better off overlaying multiple commit log streams of a shard, including those of schema changes and different raft groups the shard owns, over a single segment stream of a shard. In other words, it should be possible to put multiple logical or virtual commit logs into a single physical commit log rather than create an own segment manager for each commit log. The main reason is that we should be able to, at some point, build a total order of all changes in all commit logs. It pains me to think what it would take to consistently checkpoint (rotate) many independent log streams e.g. for the purposes of a consistent backup. In order to recover some entries ahead of the others we could index these entries in any way we wish, including chaining entries of the same "virtual" commit log within the physical log using pointers, or building a side index structure.

kostja commented:
The very requirement of recovering the schema changes fully ahead of all entries that depend on them is coming from the eventually consistent environment and won't be necessary if we implement fencing for all writes, thus ensuring that no write enters the commit log if the dependent schema is not in it. Overall it feels we're torn trying to support requirements from both universes, the strongly consistent one and the eventual one.

avikivity commented:
> The very requirement of recovering the schema changes fully ahead of all entries that depend on them is coming from the eventually consistent environment and won't be necessary if we implement fencing for all writes, thus ensuring that no write enters the commit log if the dependent schema is not in it. Overall it feels we're torn trying to support requirements from both universes, the strongly consistent one and the eventual one.

The commitlog for a data mutation can be on a different shard than the commitlog for a schema mutation. So this fencing requires markers in the commitlog (so a commitlog in shard A can wait for replay in shard B). Moreover, right now schema mutation commitlog replay just writes to the table, it does not trigger a schema re-read. So your proposal would require many changes to implement.

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/323/

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/368/

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/495/

tgrabiec commented:
> I would like to see dtests that demonstrate recovery after a schema change + dependent table change (in extra durable mode where with O_DSYNC things).

We already have such dtests: https://github.com/scylladb/scylla-dtest/blob/f7428b1e1aa1791ff266ba520440279f08eccddf/commitlog_test.py#L261

tgrabiec commented:
Please hold the merge, it's not ready yet.

tgrabiec commented:
Changes in V3:

  - Fixed dtest failures. Tests did not expect the new commitlog replay log messages and got confused.
    Changed the code to avoid initialization of schema commit log if it's not enabled in the cluster, this way log
    messages are the same as before the change.
 
  - Note: fresh cluster still needs a rolling restart for the feature to be enabled, because bootstrap is detected
    after database instance is initialized, and we need to make a decision before that, and we don't support hot
    switching.

avikivity commented:
> Now it's ready.
> 
> Changes in V3:
> 
> * Fixed dtest failures. Tests did not expect the new commitlog replay log messages and got confused.
>   Changed the code to avoid initialization of schema commit log if it's not enabled in the cluster, this way log
>   messages are the same as before the change.

This indicates the new code isn't tested at all by dtest.

Also, isn't it fragile? If a dtest restarts a node, it will get the new messages.


> * Note: fresh cluster still needs a rolling restart for the feature to be enabled, because bootstrap is detected
>   after database instance is initialized, and we need to make a decision before that, and we don't support hot
>   switching.

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/711/

tgrabiec commented:
Changes in V4:
   - Make sure new installations use the new schema commitlog immediately. This is achieved by adding an option to scylla.yaml to force the new commit log, true in the yaml (fresh installs inherit), defaults to false (for existing installs). This also makes sure that dtests work with new behavior by default

There are new failures in alternator unit tests, to be investigated.

kbr- commented:
This PR would greatly helps us test clusters with Raft. We're performing a lot of schema changes there.
cc @kostja

bhalevy commented:
@tgrabiec what's holding this PR back?

tgrabiec commented:
@bhalevy There are new failures in alternator unit tests that need to be investigated.

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/1124/

tgrabiec commented:
Changes in V5:
 
  - Fixed alternator unit test failures by https://github.com/scylladb/scylla/pull/10333/commits/683c7f93bacd34a9b5e56193e414a380ab3cb841 
    Keyspace drop was creating mutations of system.IndexInfo which was causing mixed commitlog domain transactions and apply failure on sanity check. This table is not used any more (it's a virtual table), so remove the mutation.

tgrabiec commented:
> Does this mutation participate in the schema digest?

No.

avikivity commented:
There's a logical conflict (in addition to the textual conflict) with per-partition rate limit. Need either to apply the per-partition rate limit to do_apply_many(), or to reject such requests. Since we won't have per-partition rate limits to the schema tables, rejecting is probably fine.

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/1179/

bhalevy commented:
@tgrabiec 

> CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/1179/

https://jenkins.scylladb.com/job/releng/job/Scylla-CI/1179/consoleText
```
[2022-07-05T11:20:50.777Z] ./replica/database.hh:1548:1: error: version control conflict marker in file
[2022-07-05T11:20:50.777Z] <<<<<<< HEAD
[2022-07-05T11:20:50.777Z] ^
[2022-07-05T11:20:50.777Z] 1 error generated.
```

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/1217/
